app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/RedirectHelper.java

```diff
@@ -11,8 +13,34 @@
 
     public static final String DEFAULT_REDIRECT_URL = "/applications";
     private static final String REDIRECT_URL_HEADER = "X-Redirect-Url";
+    private static final String REDIRECT_URL_QUERY_PARAM = "redirectUrl";
 
-    public static String getRedirectUrl(HttpHeaders httpHeaders) {
+    /**
+     * This function determines the redirect url that the browser should redirect to post-login. The priority order
+     * in which these checks will be made are:
+     * 1. Query parameters
+     * 2. Headers
+     *
+     * @param request
+     * @return
+     */
+    public static String getRedirectUrl(ServerHttpRequest request) {
+
+        MultiValueMap<String, String> queryParams = request.getQueryParams();
+        HttpHeaders httpHeaders = request.getHeaders();
+
+        if (queryParams != null && queryParams.containsKey(REDIRECT_URL_QUERY_PARAM)) {
+            String redirectUrl = queryParams.getFirst(REDIRECT_URL_QUERY_PARAM);
+            if (!(redirectUrl.startsWith("http://") || redirectUrl.startsWith("https://")) &&
+                    !StringUtils.isEmpty(httpHeaders.getOrigin())) {
+                redirectUrl = httpHeaders.getOrigin() + DEFAULT_REDIRECT_URL;
```

Comment:
If `redirectUrl` is a relative URL, we should be appending that to the origin instead of the default URL no?

```suggestion
                redirectUrl = httpHeaders.getOrigin() + (StringUtils.isEmpty(redirectUrl) ? DEFAULT_REDIRECT_URL : redirectUrl);
```